src/asyncresults.jl

```diff
@@ -0,0 +1,242 @@
+"An asynchronous PostgreSQL query"
+mutable struct AsyncResult
+    "The LibPQ.jl Connection used for the query"
+    jl_conn::Connection
+
+    "Whether this AsyncResult has locked the Connection"
+    conn_locked::Atomic{Bool}
+
+    "Keyword arguments to pass to Result on creation"
+    result_kwargs::Ref
+
+    "Task which errors or returns a LibPQ.jl Result which is created once available"
+    result_task::Task
+
+    function AsyncResult(jl_conn::Connection, conn_locked::Atomic{Bool}, result_kwargs::Ref)
+        async_result = new(jl_conn, conn_locked, result_kwargs)
+        jl_conn.async_result = async_result
+        return async_result
+    end
+end
+
+function AsyncResult(jl_conn::Connection, conn_locked::Atomic{Bool}; kwargs...)
+    AsyncResult(jl_conn, conn_locked, Ref(kwargs))
+end
+
+function AsyncResult(jl_conn::Connection; kwargs...)
+    AsyncResult(jl_conn, Atomic{Bool}(true); kwargs...)
+end
+
+function Base.show(io::IO, async_result::AsyncResult)
+    status = if isready(async_result)
+        if iserror(async_result)
+            "errored"
+        else
+            "finished"
+        end
+    else
+        "in progress"
+    end
+    print(io, typeof(async_result), " (", status, ")")
+end
+
+function _ensure_unlocked(async_result::AsyncResult)
+    was_locked = atomic_xchg!(async_result.conn_locked, false)
+    if was_locked
+        async_result.jl_conn.async_result = nothing
+        unlock(async_result.jl_conn)
+    end
+    return was_locked
+end
```

Comment:
this function is pretty wild, can it have a docstring on what it is for?

Reply:
Why do we have an `async_result.conn_locked` ?
Why do we not just use `islocked(async_result.conn)` ?

Reply:
Does it matter wether the lock is this results fault?

Reply:
Life just seems like it would be better with 1 less atomic floating around

Reply:
The `conn_locked` property exists to tell the AsyncResult whether it has the authority to perform operations on the connection which require a lock. The reason there's an atomic there is to prevent double-unlocking. 

`_ensure_unlocked` is called in exactly two places: 
1. In consume, on return
2. If consume isn't called due to an error

All it does is unlock the connection if it is currently locked by that AsyncResult.

Reply:
I do not love it.

Reply:
Stay tuned for this maybe going away